Add a benchmark task to CircleCI

ozevren opened the pull request:
This PR adds a benchmark checking task to the CircleCI. It is not required, and it does have an unreasonably high tolerance for runtime calculation, which I intend to tighten and fix as I make progress.

NOTE: I am sure I am going to get a lot of feedback on both CircleCI stuff, as well as naming. Let me know and I'll accommodate feedback.

istio-testing commented:
@ozevren: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 451e08e2d85aab8f785b4412663b3e41839bffa1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2592/istio-pilot-e2e/1337/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@ozevren: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 451e08e2d85aab8f785b4412663b3e41839bffa1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2592/istio-pilot-e2e/1337/) | `/test istio-pilot-e2e`
prow/istio-presubmit.sh | 9877b9890f0c8a109f2e0ad6be3c31dae177279b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2592/istio-presubmit/2533/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ozevren commented:
Any takers?

ozevren commented:
That diff issue keeps coming back. It looks like a CRLF issue. Updated the file, hopefully it will stick.

I've added stanzas to periodic, nightly and all. These tests use the go benchmark test mechanism. As such, they are generally on-par with our regular unit tests in terms of runtime.

ldemailly commented:
@ozevren sorry somehow didn't see this PR until now, I looked at the output on circleci:

```
#!/bin/bash -eo pipefail
mkdir -p /go/out/benchcheck
make benchcheck | tee -a /go/out/benchcheck/go-benchcheck-report.out
bin/perfcheck.sh
Found the following benchmark baseline files:
  /go/src/istio.io/istio/bin/../mixer/pkg/il/interpreter/bench.baseline


Running benchmarks in dir: /go/src/istio.io/istio/bin/../mixer/pkg/il/interpreter
Results:
BenchmarkInterpreter/ExprBench/ok_1st-32         	10000000	       196 ns/op	       0 B/op	       0 allocs/op
BenchmarkInterpreter/ExprBench/ok_2nd-32         	 2000000	       714 ns/op	      56 B/op	       4 allocs/op
BenchmarkInterpreter/ExprBench/not_found-32      	 2000000	       716 ns/op	      56 B/op	       4 allocs/op
BenchmarkInterpreter/true_&&_false-32            	10000000	       134 ns/op	       0 B/op	       0 allocs/op
BenchmarkInterpreter/true_&&_true-32             	10000000	       134 ns/op	       0 B/op	       0 allocs/op
BenchmarkInterpreter/false_&&_false-32           	10000000	       130 ns/op	       0 B/op	       0 allocs/op
```

Is that the delta from baseline, the new numbers ? can you make it clear (and if it doesn't have deltas, add them?)

otherwise sgtm

rshriram commented:
It’s not about runtime per test. It’s about additional resources. Once we hit the limit, the number of PRs that get tested in parallel will get throttled. Until circle adds support for ignoring auto cancelling running tests for old commits, let’s not add more tasks in the all workflow (the thing that runs every pr) 

Anything except all is fine. A nightly test should ideally invoke some script to upload test results.

ozevren commented:
@ldemailly I've clarified the wording a bit.
Adding deltas in the case of non-failure is a bit involved, and would take non-trivial changes to the shell script. (The script uses absence/presence of comparison check to set error level). I can add more detailed delta output later.

ozevren commented:
@rshriram sorry, still trying to understand how the resource limitation work. Does this mean we simply cannot add any more tasks at all without degrading parallelism?

The main value add of this gate is when it is run as part of a checkin gate. Post-facto analysis and remediation (i.e. nightly runs) is onerous on the wrong set of people (i.e. not the original submitter) and generally does not work in my experience.

I don't mind removing the all stanza for this check-in and can add it once any transient issues are resolved, but I am worried about such limitations that are imposed by CircleCI if they will hamper the work for the long term.

rshriram commented:
Why don’t you combine this with mixer tests?

ozevren commented:
@rshriram That would work. I can add it to mixer-test, but I think it is generally applicable and useful.

@ldemailly, @costinm, any concerns with adding this perf benchmark check as part of "test" (or "mixer-test") target in the Makefile?

costinm commented:
What's the extra run time ? We want 'test' target to be fast, perf doesn't
seem like a good fit.

Also would want to see it running and stable for few weeks before adding it
to any blocking target.

On Wed, Jan 24, 2018 at 8:46 AM, Ozben Evren <notifications@github.com>
wrote:

> @rshriram <https://github.com/rshriram> That would work. I can add it to
> mixer-test, but I think it is generally applicable and useful.
>
> @ldemailly <https://github.com/ldemailly>, @costinm
> <https://github.com/costinm>, any concerns with adding this perf
> benchmark check as part of "test" (or "mixer-test") target in the Makefile?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2592#issuecomment-360197227>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFI6lRyrywrC9Su3-jk5_DC1pkWrw75ks5tN15qgaJpZM4Rc592>
> .
>

ozevren commented:
Extra runtime is pretty small actually. It is based on the benchmark run times, which is generally short. As an example, the IL benchmarks that I introduced here runs for around 1s per benchmark. Depending on the timing variance in the components this time goes up, as the Go benchmark infrastructure runs the tests until it deems it is getting stable numbers.

As for running it for a few weeks to stabilize, that was my intention originally with this PR. As this PR stands, it is a separate CI gate that is not required. The intention was to have this in place, ensure it was stable for a few weeks and then make it required.

So I *do* want to get this in non-blocking mode and bake it. But for that I need to have it up and running as part of check-ins. 

I can add this to the makefile in a non-blocking manner, but I somehow need to be able to query different jobs and see the failures and analyze. Happy to follow any guidance you guys can give me with this.

costinm commented:
+1 on adding it as a separate CI job for few weeks, and merge it after it's
stable.

You could add it to the 3-hour periodic build on master ( to verify it is
stable you don't need it on each PR)

Costin

On Wed, Jan 24, 2018 at 9:26 AM, Ozben Evren <notifications@github.com>
wrote:

> Extra runtime is pretty small actually. It is based on the benchmark run
> times, which is generally short. As an example, the IL benchmarks that I
> introduced here runs for around 1s per benchmark. Depending on the timing
> variance in the components this time goes up, as the Go benchmark
> infrastructure runs the tests until it deems it is getting stable numbers.
>
> As for running it for a few weeks to stabilize, that was my intention
> originally with this PR. As this PR stands, it is a separate CI gate that
> is not required. The intention was to have this in place, ensure it was
> stable for a few weeks and then make it required.
>
> So I *do* want to get this in non-blocking mode and bake it. But for that
> I need to have it up and running as part of check-ins.
>
> I can add this to the makefile in a non-blocking manner, but I somehow
> need to be able to query different jobs and see the failures and analyze.
> Happy to follow any guidance you guys can give me with this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2592#issuecomment-360209931>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFI6mmnNJuQligz_Va9ESMoAoePwgrqks5tN2ezgaJpZM4Rc592>
> .
>

ozevren commented:
So, any objections to this plan then:
- Use CircleCI, but use periodic & nightly but not all.
- Stabilize
- Once stabilized, at it to "make test".

Is this acceptable?

ldemailly commented:
sgtm, though running in parallel once stabilized seems better, we haven't been shy so far in the number of circleci parallel jobs, not sure why this one should be different

rshriram commented on .circleci/config.yml:
```diff
@@ -297,6 +297,22 @@ jobs:
       - store_test_results:
           path: /go/out/racetests
 
+  benchcheck:
```
benchmark?

ldemailly commented:
/lgtm
can you file a separate issue about possible better output of the test we discussed

rshriram commented:
Everytime we add a new circleci task, the existing PRs or older forks of istio/istio will need to be updated to use the new task. So, whatever task we add needs to justify that. It will take a week or two before everyone has the same version of config.yml.. Hence my reluctance to add it immediately to normal PR workflow immediately, and the suggestion to merge with make mixer-test so that this test gets activated for all users.
